superset/common/query_context.py

```diff
@@ -201,7 +232,18 @@ def cache_timeout(self) -> int:
             return self.datasource.database.cache_timeout
         return config["CACHE_DEFAULT_TIMEOUT"]
 
-    def cache_key(self, query_obj: QueryObject, **kwargs: Any) -> Optional[str]:
+    def cache_key(self, **extra: Any) -> str:
+        """
+        The cache key is made out of the key/values from self.cached_values, plus any
+        other key/values in `extra`
+        """
+        key_prefix = "qc-"
+        cache_dict = self.cache_values.copy()
+        cache_dict.update(extra)
+
+        return generate_cache_key(cache_dict, key_prefix)
```

Comment:
I believe this will cause unnecessary cache misses, as it's calculating the key based on the raw queries in `self.cache_values`, not `QueryObject.cache_key()`. Can we make it so that the `queries` property in `self.cache_values` is the respective `cache_key` for said `QueryObject`, not the full object?

Reply:
Not sure how this would cause cache misses, as it's separate from the `query_obj.cache_key`. This cache key is generated from the relevant `QueryContext` values and passed directly into an async event payload ([here](https://github.com/apache/incubator-superset/blob/e2dc30eab4f315f110a383b516cf70bd622496fe/superset/tasks/async_queries.py#L48)) that is consumed by the async event frontend middleware. The `QueryContext` object wasn't previously cached, so I don't think this should affect any chart data query caching, unless I'm missing something.

Reply:
It seems that we're calculating a hash on the `cache_values` dict, which contains the raw individual query objects. For the same reason that we don't calculate the cache key on the raw query dict but instead call `QueryObject.cache_key()`, we need to call the same method when calculating the cache key for `QueryContext`. So something like this:
```python
self.cache_values = {
   ...
   "queries": [QueryObject(**query_obj).cache_key() for query_obj in queries],
   ...
}
```
This ensures any changes to the `QueryObject`'s cache key logic is always reflected in the `QueryContext` cache key.

Reply:
Yep, that makes sense. I guess my point here is that we're not doing a reverse-lookup of a cached `QueryContext` object based on the contents of the object. We're only ever loading this cached object via the cache key, so the contents that determine the cache key simply need to be unique enough not to collide with other objects.

Reply:
I looked at this again, and what's being cached in the `QueryContext` object should be as close to the original `form_data` passed in from the client, as that's what's being reinstantiated [here](https://github.com/apache/incubator-superset/blob/a8607c046e8abed780a8324af555c1c4de3aaeda/superset/charts/api.py#L616). So, caching only the `QueryObject` cache keys (for example) won't work here because we ultimately need to rehydrate a `QueryContext` object in order to fetch the underlying query data.

Reply:
I see your point. I just realized that what's being cached is pretty light, as it's just the `QueryContext`, not the full response payload. This may very well be a non-issue, and we can probably get back to this later if it turns out this causes perf issues, but I would assume not.